### PR TITLE
Component support for logger

### PR DESCRIPTION
**Pull request by davidzhao**
With the ability to dynamically update log levels per component

### CODE REVIEW

File: logger/logger.go [hunk 1]
```diff
@@ -239,6 +309,10 @@ func (l *ZapLogger) WithoutSampler() Logger {
 	return &dup
 }
 
+func (l *ZapLogger) isEnabled(level zapcore.Level) bool {
+	return l.sharedConfig.isEnabled(l.component, level)
```
Comment (boks1971) [inline logger/logger.go:313]: I am possibly not understanding this correctly. Is it possible to update loggers with component level if the logger has a component. This one is doing a sync map load on every log line. Feels like that could be expensive. Even if not expensive, feels like this logger struct can store the level based on component and just do a simple level check?

**Review by boks1971 — COMMENTED** [review]
Wanted to bring up the issue of hierarchical loggers. Is it possible to support that? For example, `rtc` is one level down, but there could be `rtc.room`, `rtc.participant` etc. So, we can just change log level of any depth in the hierarchy. If `rtc` is set, applies to all loggers in that namespace.

Guess that makes this more complex, but would be nice to have that granularity. If it is not too much effort, probably worth doing it now. If not, maybe v2 then?

File: logger/logger.go (hunk 1 above)
Comment (paulwe) [reply logger/logger.go:313]: we could probably load the log levels into something like `map[component]zap.AtomicLogLevel` then load the `zap.AtomicLogLevel` once when the logger is created and use `(zap.AtomicLogLevel).SetLevel` when the config updates. then we wouldn't need the config callback thing or custom level filtering. 

any log levels from the map unspecified by the updated config would revert to the default level. eliminating the callback list also removes the risk of a memory leak when creating these loggers outside of component init.

**Review by paulwe — APPROVED** [review]

**Comment by paulwe** [comment]
log level cascading for hierarchical loggers could be applied once when the config is translated to the map of atomic levels or during `WithComponent` for unspecified subcomponents... this could be implemented without adding a meaningful amount of complexity.

**Comment by boks1971** [comment]
> log level cascading for hierarchical loggers could be applied once when the config is translated to the map of atomic levels or during `WithComponent` for unspecified subcomponents... this could be implemented without adding a meaningful amount of complexity.

Yeah agreed, that makes sense. if we parse config and apply to loggers on config change, hierarchical  should be straightforward.

**Comment by davidzhao** [comment]
Thanks for the feedback guys! made this hierarchical and using AtomicLevel as @paulwe suggested.

**Comment by paulwe** [comment]
sweet, lgtm! 🚀

**Pull request merged**